contracts/Dash.sol

```diff
@@ -0,0 +1,153 @@
+
+    struct ProductItem {
+        string productId;
+        uint price;
+    }
+
+    AggregatorV3Interface internal priceAggregator;
+    bool blocked;
+    address aggregatorAddress;
+    address paymentDestination;
+    address rewardToken;
+    uint rewardAmount;
+    mapping(string => uint) private productPrices;
+    mapping(address => address) private owners;
+    mapping(address => string) private data;
+
+    modifier onlyOwners() {
+        require(owners[msg.sender] == msg.sender, "FORBIDDEN");
+        _;
+    }
+
+    uint private unlocked = 1;
+    modifier lock() {
+        require(unlocked == 1, "LOCKED");
+        unlocked = 0;
+        _;
+        unlocked = 1;
+    }
+
+    modifier notEmpty(string memory _value) {
+        bytes memory byteValue = bytes(_value);
+        require(byteValue.length != 0, "NO_VALUE");
+        _;
+    }
+
+    event Payment(address from, string productId, uint paymentAmount, uint nativeCoinPrice);
+
+    constructor(
+        address _owner,
+        address _paymentDest,
+        address _rewardToken,
+        uint _rewardAmount,
+        address _aggregatorAddress
+    ) {
+        aggregatorAddress = _aggregatorAddress;
+        priceAggregator = AggregatorV3Interface(_aggregatorAddress);
+        owners[_owner] = _owner;
+        paymentDestination = _paymentDest;
+        rewardToken = _rewardToken;
+        rewardAmount = _rewardAmount;
+    }
+
+    function getLatestPrice() public view returns (uint) {
+        (,int price,,,) = priceAggregator.latestRoundData();
+        if (price >= 0) {
+            return uint(price);
+        }
+        return 0;
+    }
+
+    function getData(address _target) public view returns(string memory) {
+        return data[_target];
+    }
+
+    function payment(
+        string memory _productId,
+        string memory _data,
+        address _rewardReceiver
+    ) external payable lock notEmpty(_productId) {
+        if (blocked) revert('BLOCKED');
+        require(aggregatorAddress != address(0), "NO_AGGREGATOR_ADDR");
+        uint nativeCoinPrice = getLatestPrice();
+        uint productPrice = productPrices[_productId];
+        require(nativeCoinPrice > 0 && productPrice >= 0, "WRONG_PRICE");
+        while(productPrice < nativeCoinPrice) {
+            productPrice = productPrice.mul(10);
+        }
+        uint weiDecimals = 10 ** 18;
+        uint amount = productPrice.mul(weiDecimals) / nativeCoinPrice;
+        uint acceptableInaccuracy = amount / 100; // 1%
+        require(msg.value > (amount - acceptableInaccuracy) && msg.value < (amount + acceptableInaccuracy), "WRONG_VALUE");
+        (bool success,) = paymentDestination.call{ value: msg.value }("");
+        require(success, "FAILED_PAYMENT");
+        _setData(msg.sender, _data);
+        emit Payment(msg.sender, _productId, msg.value, nativeCoinPrice);
+        if (IERC20(rewardToken).balanceOf(address(this)) >= rewardAmount) {
+            sendReward(_rewardReceiver);
+        }
```

Comment:
payment method flow. Start params:
- productPrice = 935$
- nativeCoinPrice = 37096522781 (assume we're on BSC. It's a BNB price: 370.96522781 (37096522781 / 10 ** 8))
```solidity
// oractle price for the native coin
// * if we want to see the real price, need to divide it by 10 ** 8
uint nativeCoinPrice = getLatestPrice(); 
// USD price
uint productPrice = productPrices[_productId];

while(productPrice < nativeCoinPrice) {
    productPrice = productPrice.mul(10);
}
// now product price is 93500000000

uint weiDecimals = 10 ** 18;
uint amount = productPrice.mul(weiDecimals) / nativeCoinPrice;

// amount is 2520451864234795000
// assume my msg.value is 2510000000000000000 (2.51 BNB)
// in case if front side have some differentce in the prices we add this percent
uint acceptableInaccuracy = amount / 100; // 1%
// acceptableInaccuracy is 25204518642347950

//    2510000000000000000 > 2495247345592447000  &&  2510000000000000000 < 2545656382877143000
require(msg.value > (amount - acceptableInaccuracy) && msg.value < (amount + acceptableInaccuracy), "WRONG_VALUE");

(bool success,) = paymentDestination.call{ value: msg.value }("");
require(success, "FAILED_PAYMENT");
setData(_key, _data);
emit Payment(msg.sender, _productId, msg.value, nativeCoinPrice);
if (IERC20(rewardToken).balanceOf(address(this)) >= rewardAmount) {
    sendReward(_rewardReceiver);
}
```